lib/synchronised_migration/main.rb

```diff
@@ -15,24 +15,43 @@ class << self
   end
 
   def call
-    lock_and_execute
+    done_or_execute
   end
 
   private
 
+  def done_or_execute
+    return Result.ok if previous_success?
+    result = lock_and_execute
+    mark_successful if result.success?
+    result
+  end
+
   def lock_and_execute
     redlock.lock! lock_key, timeout do
       execute
     end
   end
 
   def execute
-    return Result.new 'Halting the script because the previous migration failed.' if previous_failed?
+    return Result.fail 'Halting the script because the previous migration failed.' if previous_failed?
     mark_failed
     migrate
-    return Result.new 'Migration failed.' if migration_failed?
+    return Result.fail 'Migration failed.' if migration_failed?
     remove_fail_marker
-    Result.new
+    return Result.ok
+  end
+
+  def previous_success?
```

Comment:
I was having difficulty mapping this method to the description of what it's doing and I think it is related to its name. This is returning success status for a given version yeah? If so, it's probably better if it is named after that use case.